Alerting: Refactor migration to return pairs of legacy and upgraded structs

JacobsonMT opened the pull request:
Some refactoring that will simplify next changes for dry-run PRs. This should be no-op as far as the created ngalert resources and database state, though it does change some logs.

The key change here is to modify `migrateOrg` to return pairs of legacy struct + ngalert struct instead of actually persisting the alerts and alertmanager config. This will allow us to capture error information during dry-run migration.

It also moves most persistence-related operations such as title deduplication and folder creation to the right before we persist. This will simplify eventual partial migrations (individual alerts, dashboards, channels, ...).

Additionally it changes channel code to deal with `PostableGrafanaReceiver` instead of `PostableApiReceiver` (integration instead of contact point).

JacobsonMT committed:
No-op refactoring to simplify future changes

JacobsonMT committed:
Linting

yuri-tceretian commented on pkg/services/ngalert/migration/channel.go:
```diff
@@ -26,68 +24,42 @@ const (
-		empty = false
 		route, err := createRoute(c, receiver.Name)
 		if err != nil {
+			om.log.Warn("Failed to create route", "type", c.Type, "name", c.Name, "uid", c.UID, "error", err)
```
Do we need warn here? It returns error and will (should) be logged upstream.

JacobsonMT replied:
True, this is mostly to get logs for discontinued channels which do not cause an error in the current version, I'll put it in the if statement and remove the log on the route creation.

yuri-tceretian commented on pkg/services/ngalert/migration/channel.go:
```diff
@@ -26,68 +24,42 @@ const (
-
-	return nil
-}
+var ErrDiscontinued = errors.New("discontinued")
```
I think it should be at the top of the file.

yuri-tceretian commented on pkg/services/ngalert/migration/silences.go:
```diff
@@ -92,16 +92,15 @@ func (om *OrgMigration) addNoDataSilence(rule *models.AlertRule) error {
 	}
 
-	f, err := openReplace(silencesFileNameForOrg(om.cfg.DataPath, om.orgID))
+	f, err := openReplace(silencesFileNameForOrg(dataPath, orgID))
```
should not we do `defer f.Close()` here?

JacobsonMT replied:
The close method on `replaceFile` does more than just close the file. The lack of defer feels wrong but seems intentional the way this is written, sadly without any comment explaining why. I'll give it a try but am leaning toward not touching this code further if possible.

yuri-tceretian replied:
Since that's been around since long ago, I think we can skip it for now.

yuri-tceretian commented on pkg/services/ngalert/migration/persist.go:
```diff
@@ -0,0 +1,182 @@
+	}
+
+	return nil
+}
```
This file contains pieces moved from other files. Will it be hard to move tests too?

JacobsonMT replied:
I agree, but it will make rebasing a bit crazy to do this right now. I'll improve this later on.

yuri-tceretian commented on pkg/services/ngalert/migration/persist.go:
```diff
@@ -0,0 +1,182 @@
+)
+
+// Sync is a helper struct for persisting migration changes to the database.
+type Sync struct {
```
do we need to make it public?

JacobsonMT replied:
No, it doesn't need to be public right now. Holdover from a different version where it was in its own package. I'll remove it for now.

yuri-tceretian commented on pkg/services/ngalert/migration/permissions.go:
```diff
@@ -94,7 +82,7 @@ func (om *OrgMigration) getOrCreateMigratedFolder(ctx context.Context, l log.Log
 	// The general alerting folder is only accessible to admins.
 	if parentFolder == nil {
-		l.Warn("Migrating alert to the general alerting folder: original folder not found")
+		l.Debug("Migrating alert to the general alerting folder")
```
I think it is worth to be at least info. WDYT?

yuri-tceretian reviewed (approved):
LGTM. Just a few questions.

I like the direction of reducing the scope of some models and getting rid of some unnecessary wrappers.

JacobsonMT committed:
Address PR comments

JacobsonMT committed:
Move title deduplication and folder creation to pre-persist

Also refactors main migration logic to allow collection of error information.
This is currently not used but will be in dry-run.

JacobsonMT committed:
Address PR comments

JacobsonMT commented:
/deploy-to-hg

JacobsonMT committed:
linting

yuri-tceretian commented on pkg/services/ngalert/migration/ualert.go:
```diff
@@ -2,109 +2,120 @@ package migration
+		if err != nil && om.failOnError {
+			return nil, fmt.Errorf("migrate dashboard '%d': %w", dashID, err)
 		}
+		dashboardUpgrades = append(dashboardUpgrades, du)
```
This is counterintuitive because err != nil path is not always terminated. THere needs to be either a comment explaining a situation when we would like to add `du` to the result when `err != nil` or error handling piece rewritten to terminate loop in case of error.

JacobsonMT replied:
Yah, this is not right. I ended up refactoring towards what we discussed: removing failOnError, collecting the errors and failing at the end if any errors exist.

https://github.com/grafana/grafana/pull/79719/files#diff-0f24ee88a6a53f55b03ad33be23aea4d8674537474027be555073431d6c8d34bR181-R198

yuri-tceretian commented on pkg/services/ngalert/migration/alert_rule.go:
```diff
@@ -9,29 +9,27 @@ import (
 	lbls[ngmodels.MigratedUseLegacyChannelsLabel] = "true"
-	for _, c := range channels {
-		lbls[contactLabel(c.Name)] = "true"
-	}
```
It looks like the MigratedUseLegacyChannelsLabel above is also related to channels. I think it makes sense to move it too.

JacobsonMT replied:
Adding the `MigratedUseLegacyChannelsLabel` label is not conditional on anything, it always needs to be there so that default legacy notification channels (using `alertname~=.+`) will match. Given that, I slightly lean toward leaving it here but don't feel strongly about it.

yuri-tceretian commented on pkg/services/ngalert/migration/persist.go:
```diff
@@ -0,0 +1,324 @@
+}
+
+// handleAddRules persists the given add rule delta to the state and database.
+func (sync *sync) handleAddRules(ctx context.Context, state *migrationStore.OrgMigrationState, delta StateDelta, amConfig *migmodels.Alertmanager) error {
```
amConfig is not used here. Do you plan to use it in future PRs?

yuri-tceretian commented on pkg/services/ngalert/migration/persist.go:
```diff
@@ -0,0 +1,324 @@
+}
+
+// attachContactPointLabels attaches contact point labels to the given alert rules.
+func (sync *sync) attachContactPointLabels(ctx context.Context, pairs []*migmodels.AlertPair, amConfig *migmodels.Alertmanager) ([]models.AlertRule, error) {
```
amConfig is not used here. Do you plan to use it in future PRs?

yuri-tceretian commented on pkg/services/ngalert/migration/persist.go:
```diff
@@ -0,0 +1,324 @@
+	contactPairs []*migmodels.ContactPair,
+) error {
+	delta, err := createDelta(dashboardUpgrades, contactPairs)
+	if err != nil {
```
It never returns error. Do we need it?

yuri-tceretian commented on pkg/services/ngalert/migration/persist.go:
```diff
@@ -0,0 +1,324 @@
+	}
+
+	for _, pair := range delta.ChannelsToAdd {
+		if pair.ContactPoint != nil && pair.Route != nil {
```
My first question was "what if only one is specified, and why we skip it?" Digging through the changes I found that it seems to be only possible when field Error is not empty. I wonder if we can make it more clear here by 
1. checking if Error is empty (or create a method IsValid)
2. handle ContactPoint and Route separately, ie `if pair.ContactPoint != nil { add receiver }; if pair.Route != nil { add route }.
I think this will be more straightforward.

JacobsonMT replied:
Yes, this is a bit awkward because we're collecting errors but not doing anything with them yet. This will change in the PR that starts using them in a similar way to what you're saying.

yuri-tceretian commented on pkg/services/ngalert/migration/persist.go:
```diff
@@ -0,0 +1,324 @@
+func (sync *sync) handleAddRules(ctx context.Context, state *migrationStore.OrgMigrationState, delta StateDelta, amConfig *migmodels.Alertmanager) error {
+	pairs := make([]*migmodels.AlertPair, 0)
+	createdFolderUIDs := make(map[string]struct{})
+	if len(delta.DashboardsToAdd) > 0 {
```
I think we can remove it because for-loop does pretty much the same

yuri-tceretian commented on pkg/services/ngalert/migration/models.go:
```diff
@@ -22,69 +21,77 @@ type OrgMigration struct {
+		}
+	}
+
+	channel, err := c.fetch(ctx, key)
```
should we fetch when `UID = "" and ID == 0`?

JacobsonMT replied:
Legacy notifications without an ID or a UID should not be possible, but you're right this is a regression. To keep parity we should skip these malformed entries instead of throw an error. I'll make an explicit check in `extractChannels`

JacobsonMT committed:
Address PR comments
